qiskit/transpiler/passes/utils/block_to_matrix.py

```diff
@@ -0,0 +1,59 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Converts any block of 2 qubit gates into a matrix."""
+
+from numpy import identity, kron
+from qiskit.circuit.library import SwapGate
+from qiskit.quantum_info import Operator
+from qiskit.exceptions import QiskitError
+
+
+SWAP_GATE = SwapGate()
+SWAP_MATRIX = SWAP_GATE.to_matrix()
+IDENTITY = identity(2, dtype=complex)
+
+
+def _block_to_matrix(block, block_index_map):
+    """
+    The function converts any sequence of operations between two qubits into a matrix
+    that can be utilized to create a gate or a unitary.
+
+    Args:
+        block (List(DAGOpNode)): A block of operations on two qubits.
+        block_index_map (dict(Qubit, int)): The mapping of the qubit indices in the main circuit.
+
+    Returns:
+        NDArray: Matrix representation of the block of operations.
+    """
+    matrix = identity(2 ** len(block_index_map), dtype=complex)
```

Comment:
An exception should be raised if the dimension of matrix != 4.

Reply:
For this, it would be better to just throw an exception whenever `len(block_index_map) != 2` . Since earlier in the code we know that `block_index_map` comes from the following operation:

https://github.com/Qiskit/qiskit-terra/blob/c87d00c90af0635561c97a8ae54cd7eeccf78fce/qiskit/transpiler/passes/optimization/consolidate_blocks.py#L106

Which takes `block_qargs` (The set of all qubits in the block) as an argument and returns a mapped dictionary of the same length. Before we call `_block_to_matrix` we will check that `len(block_qargs) <= 2` to call it, otherwise, the old method will be used. As shown here:

https://github.com/Qiskit/qiskit-terra/blob/c87d00c90af0635561c97a8ae54cd7eeccf78fce/qiskit/transpiler/passes/optimization/consolidate_blocks.py#L112-L123

So it would be better to just add an exception at the beginning of `block_to_matrix` that says if the length of `block_index_length` exceeds 2 then throw an error. Which would cover for cases in which the matrix has a dimension that exceeds 2x2.